### PR TITLE
CORS-2814: Add support for in-cluster DNS on Cloud Platforms when cloud DNS cannot be used

### PR DESCRIPTION
**Pull request by sadasu**
**- What I did**
Based on enhancement : https://github.com/openshift/enhancements/pull/1468

**- How to verify it**
Verification still in-progress.

**- Description for the changelog**
These changes allow MCO to read an optional configmap containing the LB IPs of the LBs configured on cloud platforms, to be used to generate in-cluster DNS for API and API-Int URLs.

Contains implementation for [CORS-2814](https://issues.redhat.com//browse/CORS-2814), [CORS-2815](https://issues.redhat.com//browse/CORS-2815) and [CORS-3169](https://issues.redhat.com//browse/CORS-3169)

### CODE REVIEW

File: pkg/operator/bootstrap.go
Comment (barbacbd) [inline pkg/operator/bootstrap.go:376]: @sadasu Remember that these will be a comma separated list of strings.
```suggestion
func readLBIPsFromConfigMap(fileData []byte) ([]string, []string, error) {
```

File: manifests/cloud-platform-alt-dns/coredns-corefile.tmpl
Comment (cybertron) [inline manifests/cloud-platform-alt-dns/coredns-corefile.tmpl:1]: A couple of thoughts on this:

- I don't think the cluster will complete installation without the ingress record. 
- Since it's likely this will need to support dual stack at some point it might not hurt to just include the [logic for the secondary record](https://github.com/openshift/machine-config-operator/blob/f41e20d6ba3951ac9ce9411f0511558c34e60e53/templates/common/on-prem/files/coredns-corefile.yaml#L21) now.

Comment (sadasu) [reply manifests/cloud-platform-alt-dns/coredns-corefile.tmpl:1]: > A couple of thoughts on this:
> 
> * I don't think the cluster will complete installation without the ingress record.

For this feature, we are lucky to be able to do this in stages. In the current stage, we continue to allow the Ingress operator to start the LB service that creates the Ingress LBs and configures the cloud DNS for *.apps. 
The in-cluster CoreDNS pod only has to be able to resolve api and api-int URLs. In the next stage, we instruct the Ingress operator to only configure the LBs and skip the configuration of the cloud DNS. We use the ingress LB IPs to update the CoreDNS pod to resolve *.apps and  This gives us the ability to test our wok in smaller increments.

Comment (sadasu) [reply manifests/cloud-platform-alt-dns/coredns-corefile.tmpl:1]: @cybertron I think both your concerns have been addressed.

**Comment by sadasu** [comment]
@cybertron, @mkowalski could PTAL ?

**Comment by sadasu** [comment]
Failures in e2e-gcp-op-single-node don't seem to be related to this PR. The failure rate on this seems high.

File: templates/common/cloud-platform-alt-dns/files/coredns.yaml
Comment (cybertron) [inline templates/common/cloud-platform-alt-dns/files/coredns.yaml:9]: I would recommend not using a platform-specific namespace. If I had it to do over again I'd have used openshift-on-prem-infra (or something like that) for all of the on-prem pods. The separate names just clutter up the namespaces and don't add any benefit.

Also, whatever you do use for the namespace it will need to be added to https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_00_namespace.yaml (unless I missed where it was being added).

File: templates/common/cloud-platform-alt-dns/files/coredns.yaml
Comment (sadasu) [inline templates/common/cloud-platform-alt-dns/files/coredns.yaml:13]: @cybertron is the above comment regarding the name also applicable for the label "app"?

File: templates/common/cloud-platform-alt-dns/files/coredns.yaml
Comment (sadasu) [reply templates/common/cloud-platform-alt-dns/files/coredns.yaml:9]: That is great advice. I am using the name "openshift-cloud-platform-infra". Let me know if that works.

File: manifests/cloud-platform-alt-dns/coredns.yaml
Comment (cybertron) [inline manifests/cloud-platform-alt-dns/coredns.yaml:6]: I don't think the namespace here actually matters (it only affects whether mirror pods show up in the API, which doesn't matter for bootstrap), but I would guess you want the new name instead.

File: templates/common/cloud-platform-alt-dns/files/coredns.yaml
Comment (cybertron) [reply templates/common/cloud-platform-alt-dns/files/coredns.yaml:13]: I don't know if this matters. I'm not positive we ever actually look at this label.

However, I think if you give these all static names you could likely get rid of the cloudPlatformShortName entirely. That wouldn't need to block this though, it could easily be done as a followup.

**Review by cybertron — APPROVED** [review]
/lgtm

I don't think either of my inline comments need to block this as neither really affects the functionality of the change. They could both be cleaned up in a followup (or not, since I don't think they matter).

File: manifests/controllerconfig.crd.yaml
Comment (sinnykumari) [inline manifests/controllerconfig.crd.yaml:280]: controllerconfig CRD lives now in openshift/api , you will need to update https://github.com/openshift/api/tree/master/machineconfiguration/v1 and vendor latest changes here.

Comment (sadasu) [reply manifests/controllerconfig.crd.yaml:280]: @sinnykumari ControllerConfig in openshift/api has already been updated: https://github.com/openshift/api/blob/master/machineconfiguration/v1/0000_80_controllerconfig-TechPreviewNoUpgrade.crd.yaml#L1629 and vendored https://github.com/openshift/machine-config-operator/blob/master/vendor/github.com/openshift/api/machineconfiguration/v1/0000_80_controllerconfig-TechPreviewNoUpgrade.crd.yaml#L1064

I ran `make update` on my mco branch and that resulted in this manifest being updated. It appears that it is unnecessary. So, I will remove this from my commit.

Comment (sinnykumari) [reply manifests/controllerconfig.crd.yaml:280]: ah ok. @cdoern is this expected?

**Review by cybertron — COMMENTED** [review]
/lgtm

File: manifests/controllerconfig.crd.yaml
Comment (sadasu) [reply manifests/controllerconfig.crd.yaml]: @cdoern the same manifest is updated when I run `make update` on master. So, nothing specific to this PR. Removed this from my commit.

Comment (cdoern) [reply manifests/controllerconfig.crd.yaml]: This is expected. Make update brings in the new manifests and copies them where necessary. There is a hack script that lists all of the  CRDs we copy.

**Review by sinnykumari — COMMENTED** [review]
Overall this looks fine to me. 
I don't have much expertise in technical details going on here, I assume that @cybertron has taken care of that part.
Before adding approval, few sanity check questions:
- Linked enhancement https://github.com/openshift/enhancements/pull/1468 got closed. Is this design already been agreed upon with needed people? 
- Do we need review from additional domain expert people?
- We usually do pre-merge testing of feature. Are you working with any QE for verifying this feature or this will get tested later by corresponding QE?

**Review by mkowalski — CHANGES_REQUESTED** [review]
As per OCPBUGS-26951, if you are adding new CoreDNS pods you should also annotate them for the workload partitioning. Look at https://github.com/openshift/machine-config-operator/pull/4143/files to see how it's done for on-prem platforms

**Comment by sadasu** [comment]
> Overall this looks fine to me. I don't have much expertise in technical details going on here, I assume that @cybertron has taken care of that part. Before adding approval, few sanity check questions:
> 
> * Linked enhancement [CORS-2062: Customer configured DNS for cloud platforms AWS, Azure and GCP enhancements#1468](https://github.com/openshift/enhancements/pull/1468) got closed. Is this design already been agreed upon with needed people?
The design has been agreed upon. The enhancement has to be updated. Implementation took priority over that.

> * Do we need review from additional domain expert people?
I think the experts on on-prem CoreDNS are already involved.

> * We usually do pre-merge testing of feature. Are you working with any QE for verifying this feature or this will get tested later by corresponding QE?
QE is testing this feature on 4.15. Hence, the backport PR https://github.com/openshift/machine-config-operator/pull/4155 has been created. I am working with @yunjiang29.

File: templates/common/cloud-platform-alt-dns/files/coredns.yaml
Comment (mkowalski) [inline templates/common/cloud-platform-alt-dns/files/coredns.yaml:7]: This manifest here deserves the `target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'` annotation I mentioned above. You have it in `manifests/` but not in `templates/`, I guess that was a simple oversight

Comment (sadasu) [reply templates/common/cloud-platform-alt-dns/files/coredns.yaml:7]: Now, both of them have it. Thanks!

**Comment by sinnykumari** [comment]
Thanks, this seems to be a feature at the moment. Features are usually not backported, you may require staff-eng approval to that.
/approve

**Pull request merged**
